src/main/scala/is/hail/stats/RegressionUtils.scala

```diff
@@ -287,92 +287,83 @@ object RegressionUtils {
     }
   }
 
-  // constructs DenseVector of dosages (with missing values mean-imputed) in parallel with other statistics sufficient for linear regression
-  def toLinregDosageStats(gs: Iterable[Genotype], y: DenseVector[Double], mask: Array[Boolean], minAC: Int): Option[(DenseVector[Double], Double, Double)] = {
-    val nMaskedSamples = y.length
-    val valsX = Array.ofDim[Double](nMaskedSamples)
+  def dosageStats(gs: Iterable[Genotype], optMask: Option[Array[Boolean]], nKept: Int): (DenseVector[Double], Double) = {
+    val valsX = Array.ofDim[Double](nKept)
+    val missingRowIndices = new ArrayBuilder[Int]()
     var sumX = 0d
-    var sumXX = 0d
-    var sumXY = 0d
-    var sumYMissing = 0d
     var nMissing = 0
-    val missingRowIndices = new ArrayBuilder[Int]()
-
-    val gts = gs.dosageIterator
-    var i = 0
     var row = 0
-    while (gts.hasNext) {
-      val gt = gts.next()
-      if (mask(i)) {
-        if (gt != -1) {
-          valsX(row) = gt
-          sumX += gt
-          sumXX += gt * gt
-          sumXY += gt * y(row)
-        } else {
-          nMissing += 1
-          missingRowIndices += row
-          sumYMissing += y(row)
-        }
-        row += 1
-      }
-      i += 1
-    }
-
-    val nPresent = nMaskedSamples - nMissing
-
-    if (sumX < minAC)
-      None
-    else {
-      val meanX = sumX / nPresent
-      val missingRowIndicesArray = missingRowIndices.result()
-
-      var i = 0
-      while (i < missingRowIndicesArray.length) {
-        valsX(missingRowIndicesArray(i)) = meanX
-        i += 1
-      }
-
-      val x = DenseVector(valsX)
-      val xx = sumXX + meanX * meanX * nMissing
-      val xy = sumXY + meanX * sumYMissing
+    val gts = gs.dosageIterator
 
-      Some(x, xx, xy)
+    optMask match {
+      case Some(mask) =>
```

Comment:
That's a related but separate set of changes. This arc is toward "de-optimizing" what was over-optimized to simplify the code base (remember we discussed the problem of quadratic growth in RegUtils?).  Plenty more code to delete.